### PR TITLE
ENH: higher-dimensional input in `spatial.SphericalVoronoi`

### PR DESCRIPTION
**Pull request by pmla**
#### What does this implement/fix?

This adds support for higher-dimensional input in `spatial.SphericalVoronoi`.  ~It also adds a `circumradii` attribute, which calculates the radius of each facet.~

#### Additional information

Builds on the discussion in #10494

### CODE REVIEW

File: scipy/spatial/_spherical_voronoi.py [hunk 1]
```diff
@@ -53,17 +41,20 @@ class SphericalVoronoi:
...
         center of the sphere
-    vertices : double array of shape (nvertices, 3)
+    vertices : double array of shape (nvertices, ndim)
         Voronoi vertices corresponding to points
     regions : list of list of integers of shape (npoints, _ )
         the n-th entry is a list consisting of the indices
         of the vertices belonging to the n-th point in points
+    circumradii : double array of shape (nvertices,)
+        The radius of the spherical caps which cover each facet.  This is also
+        the distance from a Voronoi vertex to its nearest input points.
```
Comment (tylerjereddy) [inline scipy/spatial/_spherical_voronoi.py]: This sounds a lot like the relationship between Delaunay and Voronoi. I find the usage of the circumscribed circle here a little confusing relative to the case in the Delaunay modification to produce its dual (Voronoi).

**Review by tylerjereddy — COMMENTED** [review]
I'll need a bit more time to think about this one--I put a few initial comments/ questions. @rtshort may want to take a look as well.

File: scipy/spatial/_spherical_voronoi.py [hunk 2]
```diff
@@ -238,5 +227,6 @@ def sort_vertices_of_regions(self):
         generator in points and obtain a sorted version of the vertices
         of its surrounding region.
         """
-
+        if self.dim != 3:
+            raise Exception("Only supported for three-dimensional point sets")
```
[the remark this answers is not in the document]
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: Good suggestion.  I have replaced it with a `TypeError` and added a unit test for it.

File: scipy/spatial/_spherical_voronoi.py [hunk 3]
```diff
@@ -141,33 +129,31 @@ class SphericalVoronoi:
 
     """
 
-    def __init__(self, points, radius=1, center=(0, 0, 0), threshold=1e-06):
+    def __init__(self, points, radius=1, center=None, threshold=1e-06):
```
Comment (pmla) [inline scipy/spatial/_spherical_voronoi.py]: I had to change the default back to `center=None` :)
I can't see a way to set the default to all zeros when the input is variable length.  Any suggestions?

File: scipy/spatial/_spherical_voronoi.py [hunk 4]
```diff
@@ -53,17 +41,20 @@ class SphericalVoronoi:
...
-    center : double array of shape (3,)
+    center : double array of shape (ndim,)
         center of the sphere
-    vertices : double array of shape (nvertices, 3)
+    vertices : double array of shape (nvertices, ndim)
         Voronoi vertices corresponding to points
     regions : list of list of integers of shape (npoints, _ )
         the n-th entry is a list consisting of the indices
         of the vertices belonging to the n-th point in points
+    circumradii : double array of shape (nvertices,)
```
[the remark this answers is not in the document]
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: I have removed this for now.

File: scipy/spatial/_spherical_voronoi.py (hunk 1 above)
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: I have removed this for now.

**Comment by pmla** [comment]
Thanks for another great review @tylerjereddy - I am learning a lot about good practices.

I removed the `circumradii` attribute.  I will think about it some more and perhaps put it in a later PR.

File: scipy/spatial/_spherical_voronoi.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: I have removed this for now - see below.

**Comment by tylerjereddy** [comment]
What's the application / use case for the circumradii? I realize you've removed that now. Is that useful in the crystallography domain/ your manuscript?

**Comment by tylerjereddy** [comment]
Ah, you were talking about the Delaunay circumcircles, I think, so yeah maybe awkard for those here.

File: scipy/spatial/tests/test_spherical_voronoi.py [hunk 5]
```diff
@@ -170,3 +172,15 @@ def test_single_hemisphere_handling(self):
...
+        with assert_raises(ValueError):
+            sv = spherical_voronoi.SphericalVoronoi(np.linspace(0, 1, 10))
+
+    @pytest.mark.parametrize("dim", range(2, 6))
+    def test_higher_dimensions(self, dim):
+        n = 100
+        points = np.random.randn(n, dim)
+        points /= np.linalg.norm(points, axis=1)[:, np.newaxis]
+        sv = SphericalVoronoi(points)
+        assert sv.vertices.shape[1] == dim
```
Comment (tylerjereddy) [inline scipy/spatial/tests/test_spherical_voronoi.py]: This is a sensible check for higher dims. I'm having a hard time thinking of easy-to-understand ways to write unit tests for probing higher-dimensional spherical Voronoi results, especially given the complications for plotting > 3D to reason about what is going on.

One perhaps obvious idea is to have coordinates with i.e., [x, y, z, b], where all b values are constant--so that the 4th dimension is predictably not doing much & then compare back to a 3D result that is well understood.

File: scipy/spatial/_spherical_voronoi.py (hunk 3 above)
Comment (tylerjereddy) [reply scipy/spatial/_spherical_voronoi.py]: Not off the top of my head, and I think Pauli recommended avoiding mutable kwargs in your previous PR to adjust the API.

One thing for us to keep in mind is that if the API change / deprecation from your previous PR goes out in 1.4 and this restoration doesn't happen until later, that might be a little bit awkward from the user experience standpoint I suppose. I'll put a 1.4 milestone on this PR just so we are reminded to come back to this before the release if this PR is still open.

**Comment by rtshort** [comment]
Just FYI, I need a Voronoi diagram with both the circumcenters and the circumradii for my application.  Of course the circumcenters are the Voronoi vertices, so that part is given.  As long as the vertices are associated with the hull simplices then I can get the circumradii.

BTW, this enhancement is very cool.  I don't really need it in my world, but very cool.

File: scipy/spatial/tests/test_spherical_voronoi.py (hunk 5 above)
Comment (pmla) [reply scipy/spatial/tests/test_spherical_voronoi.py]: I have added two unit tests for the cross-polytope https://en.wikipedia.org/wiki/Cross-polytope
This is a simple test for higher-dimensional input.

It doesn't catch everything but it is a good sanity check on basic behaviour.

File: scipy/spatial/_spherical_voronoi.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tylerjereddy) [reply scipy/spatial/_spherical_voronoi.py]: `ValueError` [may be more appropriate](https://stackoverflow.com/a/48343484/2942522)--perhaps just switch that over the next time you make a more substantial change, since it is a minor point

File: scipy/spatial/_spherical_voronoi.py [hunk 6]
```diff
@@ -187,25 +173,16 @@ def _calc_vertices_regions(self):
 
         # get Convex Hull
         self._tri = scipy.spatial.ConvexHull(self.points)
-
-        # triangles will have shape: (2N-4, 3, 3)
         triangles = self._tri.points[self._tri.simplices]
 
         # get circumcenters of Convex Hull triangles from facet equations
-        # circumcenters will have shape: (2N-4, 3)
```
Comment (tylerjereddy) [inline scipy/spatial/_spherical_voronoi.py]: the comments related to space complexity are still useful in N-dimensions, right? we'd just have to generalize them

File: scipy/spatial/_spherical_voronoi.py [hunk 7]
```diff
@@ -141,33 +129,31 @@ class SphericalVoronoi:
 
     """
 
-    def __init__(self, points, radius=1, center=(0, 0, 0), threshold=1e-06):
+    def __init__(self, points, radius=1, center=None, threshold=1e-06):
+
+        self.points = np.array(points).astype(np.double)
```
Comment (tylerjereddy) [inline scipy/spatial/_spherical_voronoi.py]: How is this change related to the expanding dimensionality? The argument is already defined as an array of doubles, right? The coercion to support array-like is a separate matter, no?

File: scipy/spatial/tests/test_spherical_voronoi.py (hunk 5 above)
Comment (tylerjereddy) [reply scipy/spatial/tests/test_spherical_voronoi.py]: Here, can you also test than `len(sv.regions) == n`?

File: scipy/spatial/tests/test_spherical_voronoi.py [hunk 8]
```diff
@@ -170,3 +172,33 @@ def test_single_hemisphere_handling(self):
...
+    @pytest.mark.parametrize("dim", range(2, 7))
+    def test_higher_dimensions(self, dim):
+        n = 100
+        points = np.random.randn(n, dim)
+        points /= np.linalg.norm(points, axis=1)[:, np.newaxis]
+        sv = SphericalVoronoi(points)
+        assert sv.vertices.shape[1] == dim
+
+    @pytest.mark.parametrize("dim", range(2, 7))
+    def test_cross_polytope_regions(self, dim):
```
Comment (tylerjereddy) [inline scipy/spatial/tests/test_spherical_voronoi.py]: some brief comments to explain what these cross polytope tests are doing might be helpful--I'm still trying to wrap my head around these--I'll have to check the page you linked more closely

**Review by tylerjereddy — DISMISSED** [review]
Can you possibly expand `test_num_vertices()` to the higher dimensions as well? That will allows us to verify Euler's formula / condition that the relationship between `n` generators and `n_v` Voronoi vertices follows: `n_v = 2n - 4`. I don't know offhand if that's valid to arbitrary dims, but the citation is in the source code.

Can you point to an example publication that uses a spherical Voronoi diagram with ndim > 3? Is this the example you gave before from your crystallography work again or is there something else?

File: scipy/spatial/_spherical_voronoi.py (hunk 7 above)
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: In 3D `center=(0,0,0)` makes sense.
In 4D we need `center=(0,0,0,0)`
In 6D we need `center=(0,0,0,0,0,0)` etc.

We can't have variable length default parameters, so I had to revert to `None`.

Comment (tylerjereddy) [reply scipy/spatial/_spherical_voronoi.py]: I mean the new line with `self.points = np.array(points).astype(np.double)`.

Comment (tylerjereddy) [reply scipy/spatial/_spherical_voronoi.py]: Not saying it is a bad idea to support array-like input eventually, just saying I don't immediately see the connection with higher dimensional support.

Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: > I mean the new line with self.points = np.array(points).astype(np.double).

Ah sorry, I misunderstood.  You are right - there is no specific connection to higher-dimensional input.  If the input is an integer array though, then `self.points` inherits this type, and it is inconsistent with the attributes in the documentation.  This is perhaps a minor detail though.

If/when #10759 is merged, we will want to copy `points` rather than just having a reference, since otherwise the user could change the values of `points` before calling `calculate_areas` and get an inconsistent result.

File: scipy/spatial/_spherical_voronoi.py (hunk 6 above)
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: I couldn't figure out how to generalize Euler's formula - I don't think `n_v = 2n - 4` generalizes.  I can look into this.

**Comment by pmla** [comment]
> Can you point to an example publication that uses a spherical Voronoi diagram with ndim > 3? Is this the example you gave before from your crystallography work again or is there something else?

My motivation for extending to arbitrary dimensions comes from my crystallography work.  In general though, spherical voronoi diagrams are useful for calculating the covering radius of a point set on a hypersphere.  I will do a literature review and see what else I can find.

**Review by tylerjereddy — COMMENTED** [review]
> Theorem 6.15 (Euler). For a surface S of genus g, ξ(S) = 2 − 2g.
>
> Sketch of Proof. We aim to show that for any embedded graph G partitioning a surface S of genus g, the equation V − E + F = 2 − 2g holds. The proof is by induction on g. For g = 0, the surface S is homeomorphic to a sphere, and Theorem 6.12 applies to show that ξ(S) = 2.

Devadoss, Satyan L.; O'Rourke, Joseph. Discrete and Computational Geometry (p. 168). Princeton University Press. Kindle Edition. 

Now, it turns out that the above is for the 2-sphere (regular ball in 3D space).

To generalize even further, there's apparently [a concise proof](https://math.stackexchange.com/a/453667/480070) in `Hirsch's book "Differential Topology", page 134`, which both at that link & [on wikipedia](https://en.wikipedia.org/wiki/Euler_characteristic#Arbitrary_dimensions) produces a beautiful equation for the Euler characteristic:

> The n-dimensional sphere has Euler characteristic 1 + (−1) ^ n, for any n-sphere, so the value is always 0 or 2

So, `V - E + F = 1 + (-1) ^ 2` 

And there is confirmation we can use that for n-dimensional polytopes in [the Wolfram Reference](http://mathworld.wolfram.com/PolyhedralFormula.html), which comes from a proof by Schläfli himself.

So, we may be able to use this for unit testing of n-dimesional spherical Voronoi output. We always get the number of vertices (V) easily, but edges (E) and faces (F) in n-dimensions may be a little harder? If we're confident Qhull is checking the Euler characteristics for n-dimensional calculations we may just go with trusting that I suppose.

File: scipy/spatial/_spherical_voronoi.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: I have made it a `ValueError`.

File: scipy/spatial/_spherical_voronoi.py (hunk 6 above)
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: The generalization is described here:
https://en.wikipedia.org/wiki/Euler_characteristic#Generalizations

**Comment by pmla** [comment]
I have added tests for the Euler characteristic in arbitrary dimensions.  The generalization is described here: https://en.wikipedia.org/wiki/Euler_characteristic#Generalizations

File: scipy/spatial/_spherical_voronoi.py (hunk 6 above)
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: This relation doesn't generalize to arbitrary dimensions.

Comment (tylerjereddy) [reply scipy/spatial/_spherical_voronoi.py]: Since 3D is still likely the primary use case, perhaps still worth retaining with that clarification then

File: scipy/spatial/tests/test_spherical_voronoi.py (hunk 5 above)
Comment (pmla) [reply scipy/spatial/tests/test_spherical_voronoi.py]: Sorry - I overlooked this.  I have added the test.

File: scipy/spatial/_spherical_voronoi.py (hunk 6 above)
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: I have put the comments back in, noting that they are valid for 3D inputs.

File: scipy/spatial/tests/test_spherical_voronoi.py (hunk 8 above)
Comment (pmla) [reply scipy/spatial/tests/test_spherical_voronoi.py]: I have fixed one of the tests and added some comments.  In brief, these are polytopes whose Voronoi vertex coordinates are known, and which we can test in all dimensions.

File: scipy/spatial/_spherical_voronoi.py (hunk 7 above)
Comment (tylerjereddy) [reply scipy/spatial/_spherical_voronoi.py]: We should probably remove it then, and if there's pathological input that causes a bug perhaps deal with that in a separate PR along with a test.

File: scipy/spatial/tests/test_spherical_voronoi.py [hunk 9]
```diff
@@ -217,3 +218,67 @@ def test_geodesic_input(self, n, radius, center):
...
+    def test_higher_dimensions(self, dim):
+        n = 100
+        rng = np.random.RandomState(seed=0)
+        points = rng.randn(n, dim)
+        points /= np.linalg.norm(points, axis=1)[:, np.newaxis]
+        sv = SphericalVoronoi(points)
+        assert sv.vertices.shape[1] == dim
+        assert len(sv.regions) == n
+
+        # verify Euler characteristic
```
Comment (tylerjereddy) [inline scipy/spatial/tests/test_spherical_voronoi.py]: Good!

**Review by tylerjereddy — COMMENTED** [review]
I'd like to get this in before branching 1.4.x to avoid an API change in 1.4.x followed by a reversal of that API change in 1.5.x.

Let me resolve the merge conflict now & check a few things locally.

**Comment by tylerjereddy** [comment]
I rebased on latest master & fixed the deprecated `as_dcm()` call. I confirmed that `spherical_voronoi.py` still has 100 % line coverage in this feature branch.

Let me now check the `asv` benchmarks for 3D input.

Beyond that, I can push fixes for any of the changes you agree with---let me know.

**Comment by tylerjereddy** [comment]
I ran `asv continuous master f2f49c6d -b "SphericalVor*" >& output.txt`

There was no evidence of a slowdown in performance as a result of adding n-dimensional support, so that's good.

File: scipy/spatial/tests/test_spherical_voronoi.py [hunk 10]
```diff
@@ -183,8 +185,7 @@ def test_rank_deficient(self):
     @pytest.mark.parametrize("radius", [0.5, 1, 2])
     @pytest.mark.parametrize("center", [(0, 0, 0), (1, 2, 3)])
     def test_geodesic_input(self, n, radius, center):
-        np.random.seed(0)
-        U = Rotation.random().as_dcm()
+        U = Rotation.random(random_state=0).as_dcm()
```
[the remark this answers is not in the document]
Comment (pmla) [reply scipy/spatial/tests/test_spherical_voronoi.py]: I have moved this change into #11049

File: scipy/spatial/tests/test_spherical_voronoi.py [hunk 11]
```diff
@@ -123,6 +115,16 @@ def test_sort_vertices_of_regions_flattened(self):
...
 
+    def test_sort_vertices_of_regions_dimensionality(self):
+        with pytest.raises(TypeError, match="three-dimensional"):
+            points = np.array([[1, 0, 0, 0],
+                               [0, 1, 0, 0],
+                               [0, 0, 1, 0],
+                               [0, 0, 0, 1],
+                               [0.5, 0.5, 0.5, 0.5]])
+            sv = spherical_voronoi.SphericalVoronoi(points)
+            sv.sort_vertices_of_regions()
```
[the remark this answers is not in the document]
Comment (pmla) [reply scipy/spatial/tests/test_spherical_voronoi.py]: Thanks. I have fixed this.

File: scipy/spatial/_spherical_voronoi.py [hunk 12]
```diff
@@ -138,18 +142,18 @@ def __init__(self, points, radius=1, center=(0, 0, 0), threshold=1e-06):
...
-                          DeprecationWarning)
-
-        self.points = points
         self.radius = radius
-        self.center = np.array(center)
         self.threshold = threshold
+        self._dim = points.shape[1]
+        if center is None:
+            center = np.zeros(self._dim)
+        self.center = np.array(center)
```
[the remark this answers is not in the document]
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: I have made this change.

File: scipy/spatial/_spherical_voronoi.py [hunk 13]
```diff
@@ -128,8 +128,12 @@ class SphericalVoronoi:
     >>> plt.show()
 
     """
+    def __init__(self, points, radius=1, center=None, threshold=1e-06):
+
+        self.points = np.array(points).astype(np.float)
+        if len(points.shape) < 2 or points.shape[1] < 2:
```
[the remark this answers is not in the document]
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: I have reverted it to the original behaviour, removed the error checking and the associated test.

File: scipy/spatial/_spherical_voronoi.py (hunk 7 above)
Comment (pmla) [reply scipy/spatial/_spherical_voronoi.py]: OK. Fixed!

**Comment by pmla** [comment]
Thanks for the review @tylerjereddy.
I have made the changes you suggested. I pushed the fixes myself because my git skills are poor and this is easier for me to manage.

File: scipy/spatial/tests/test_spherical_voronoi.py [hunk 14]
```diff
@@ -123,6 +114,16 @@ def test_sort_vertices_of_regions_flattened(self):
         actual = list(itertools.chain(*sorted(sv.regions)))
         assert_array_equal(actual, expected)
 
+    def test_sort_vertices_of_regions_dimensionality(self):
+        points = np.array([[1, 0, 0, 0],
+                           [0, 1, 0, 0],
+                           [0, 0, 1, 0],
+                           [0, 0, 0, 1],
+                           [0.5, 0.5, 0.5, 0.5]])
+        with pytest.raises(TypeError, match="three-dimensional"):
```
Comment (tylerjereddy) [inline scipy/spatial/tests/test_spherical_voronoi.py]: the context for the error is still larger than it needs to be--we only need to probe `sort_vertices_of_regions()` so that we don't accidentally catch an error from `SphericalVoronoi` itself. Anyway, that is pretty minor & could be cleaned up later if at all.

**Review by tylerjereddy — APPROVED** [review]
`ppc64le` CI failure is unrelated, this looks solid now, thanks Peter

**Comment by tylerjereddy** [comment]
I updated the appropriate section of the [release notes wiki](https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.4.0), feel free to expand/correct as you see fit.
